# Contents
01 Importing Libraries
02 Fixing product price outliers
03 Creating a Dataframe for Mixed Data Type Practice
04 Missing Values
05 Finding Duplicates
06 Task Questions

# 01 Importing Libraries

In [41]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [42]:
# Folder location string
path = r'C:\Users\ckra9\Documents\CareerFoundry\Instacart Basket Analysis'

In [43]:
path

'C:\\Users\\ckra9\\Documents\\CareerFoundry\\Instacart Basket Analysis'

In [44]:
# Importing products data set from original data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [45]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

# 02 Fixing product price outliers

In [46]:
# Finding outliers in products
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [47]:
# Checking for products with abnormally high prices
df_prods[df_prods["prices"]> 50.0]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [48]:
# Replacing 2 instances of abnormally high prices with correct price.
df_prods = df_prods.replace ({"prices":{99999.0: 9.99, 14900.0:1.49 }})

In [49]:
# Importing orders data set from prepared data 
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [50]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
1                 1   2398795        1    prior             2          3   
2                 2    473747        1    prior             3          3   
3                 3   2254736        1    prior             4          4   
4                 4    431534        1    prior             5          4   
...             ...       ...      ...      ...           ...        ...   
3421078     3421078   2266710   206209    prior            10          5   
3421079     3421079   1854736   206209    prior            11          4   
3421080     3421080    626363   206209    prior            12          1   
3421081     3421081   2977660   206209    prior            13          1   
3421082     3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 8 columns]

# Exercise 4.5

In [51]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# 03 Creating a Dataframe for Mixed Data Type Practice

In [52]:
# Creating the dataframe
df_test = pd.DataFrame()

In [53]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [54]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [55]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [56]:
# Changing data type 
df_test['mix'] = df_test['mix'].astype('str')

# 04 Missing Values

In [57]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [58]:
# Creating subset of missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [59]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [60]:
df_prods.shape

(49693, 5)

In [61]:
# Creating new products data frame by erasing missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [62]:
df_prods_clean.shape

(49677, 5)

# 05 Finding Duplicates

In [63]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [64]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [65]:
# Creating new products data frame without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [66]:
df_prods_clean_no_dups.shape

(49672, 5)

In [67]:
# Exporting CSV file for cleaned products dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 06 Task Questions

In [68]:
# Importing wrangled orders 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [69]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Q2
The first thing I notice is that there is an unnamed column. I don't remember this in previous exercises. I also notice that the order_number and days_since_prior_order columns contain max amounts that are lower than their 50% and 75% percentile values.

In [70]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           1   2398795        1    prior             2          3   
2           2    473747        1    prior             3          3   
3           3   2254736        1    prior             4          4   
4           4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [71]:
# Finding mixed-type columns in orders data frame. (Q3)
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# Q4 
There are not any mixed data types within the columns of the orders dataframe.

In [72]:
# Checking orders dataframe for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [73]:
# Creating a subset of NaN values within the days_since_prior_order column
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [74]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

In [75]:
df_ords.shape

(3421083, 8)

# Q5
There are 206,209 instances of null values in the days_since_prior_order column. Every instance also has order_number = 1. This leads me to assume that there were no prior orders and this was the customer's first order.

# Q6
These missing values contain pertinent information regarding initial orders for customers. No action is recommended. 

In [76]:
# Checking orders dataframe for duplicate records. 
df_ords_dups = df_ords[df_ords.duplicated()]

In [77]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

# Q7 & 8
There are no duplicates which means no action is needed.

In [78]:
# Exporting cleaned orders dataframe as CSV file
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))